src/schemes/fluid/surface_normal_sph.jl

```diff
@@ -121,30 +133,140 @@ function remove_invalid_normals!(system::FluidSystem, surface_tension::SurfaceTe
-    (; cache, surface_normal_method) = system
+    (; cache, surface_tension) = system
 
     # Reset surface normal
     set_zero!(cache.surface_normal)
     set_zero!(cache.neighbor_count)
 
+    # TODO: if color values are set only different systems need to be called
     @trixi_timeit timer() "compute surface normal" foreach_system(semi) do neighbor_system
         u_neighbor_system = wrap_u(u_ode, neighbor_system, semi)
         v_neighbor_system = wrap_v(v_ode, neighbor_system, semi)
 
-        calc_normal_akinci!(system, neighbor_system, u, v, v_neighbor_system,
-                            u_neighbor_system, semi, surface_normal_method)
-        remove_invalid_normals!(system, surface_tension)
+        calc_normal!(system, neighbor_system, u, v, v_neighbor_system,
+                     u_neighbor_system, semi, surface_normal_method_,
+                     surface_normal_method(neighbor_system))
+    end
+    remove_invalid_normals!(system, surface_tension, surface_normal_method_)
+
+    return system
+end
+
+function calc_curvature!(system, neighbor_system, u_system, v,
+                         v_neighbor_system, u_neighbor_system, semi, surfn, nsurfn)
+end
+
+# Section 5 in Morris 2000 "Simulating surface tension with smoothed particle hydrodynamics"
+function calc_curvature!(system::FluidSystem, neighbor_system::FluidSystem, u_system, v,
+                         v_neighbor_system, u_neighbor_system, semi,
+                         surfn::ColorfieldSurfaceNormal, nsurfn::ColorfieldSurfaceNormal)
+    (; cache) = system
+    (; curvature) = cache
+
+    system_coords = current_coordinates(u_system, system)
+    neighbor_system_coords = current_coordinates(u_neighbor_system, neighbor_system)
+    nhs = get_neighborhood_search(system, neighbor_system, semi)
+    correction_factor = fill(eps(eltype(system)), n_moving_particles(system))
+
+    no_valid_neighbors = 0
+
+    foreach_point_neighbor(system, neighbor_system,
+                           system_coords, neighbor_system_coords,
+                           nhs) do particle, neighbor, pos_diff, distance
+        m_b = hydrodynamic_mass(neighbor_system, neighbor)
+        rho_b = particle_density(v_neighbor_system, neighbor_system, neighbor)
+        n_a = surface_normal(system, particle)
+        n_b = surface_normal(neighbor_system, neighbor)
+        v_b = m_b / rho_b
+
+        # eq. 22 we can test against eps() here since the surface normals that are invalid have been reset
+        if dot(n_a, n_a) > eps() && dot(n_b, n_b) > eps()
+            w = smoothing_kernel(system, distance)
+            grad_kernel = smoothing_kernel_grad(system, pos_diff, distance, particle)
+
+            for i in 1:ndims(system)
+                curvature[particle] += v_b * (n_b[i] - n_a[i]) * grad_kernel[i]
+            end
+            # eq. 24
+            correction_factor[particle] += v_b * w
+            # prevent NaNs from systems that are entirely skipped
+            no_valid_neighbors += 1
```

Comment:
Third allocation here.
I don't exactly know why, but I'm assuming this has something to do with closures. The variable is defined before the `foreach_point_neighbor` and passed together with the body as a closure, which is then causing problems.
Also, you would need a proper reduction on multiple threads.

Can't you just change the loop below to
```julia
    for i in 1:n_moving_particles(system)
        if correction_factor[i] > eps()
            curvature[i] /= correction_factor[i]
        end
    end
```